src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -103,21 +111,38 @@ private static BigDecimal[] makePower10Table() {
    * see {@link DataValidationUtil#isAllowedSemiStructuredType}.
    *
    * @param input Object to validate
-   * @return JSON tree representing the input
+   * @return Minified JSON string
    */
-  private static JsonNode validateAndParseSemiStructuredAsJsonTree(
+  private static String validateAndParseSemiStructured(
       String columnName, Object input, String snowflakeType, final long insertRowIndex) {
     if (input instanceof String) {
-      String stringInput = (String) input;
+      final String stringInput = (String) input;
       verifyValidUtf8(stringInput, columnName, snowflakeType, insertRowIndex);
-      try {
-        return objectMapper.readTree(stringInput);
-      } catch (JsonProcessingException e) {
+      final StringBuilderWriter resultWriter = new StringBuilderWriter(stringInput.length());
+      try (final JsonParser parser = factory.createParser(stringInput);
+          final JsonGenerator generator = factory.createGenerator(resultWriter)) {
+        while (parser.nextToken() != null) {
+          final JsonToken token = parser.currentToken();
+          if (token.isNumeric()) {
+            // If the current token is a number, we cannot just copy the current event because it
+            // would write token the token from double (or big decimal), whose scientific notation
+            // may have been altered during deserialization. We want to preserve the scientific
+            // notation from the user input, so we write the current numer as text.
+            generator.writeNumber(parser.getText());
```

Comment:
does datetime parsing also introduce such mutations that need to be avoided? please add tests to verify.

Reply:
This code is triggered when customers pass a string into a variant column, so we never pass dates here as the input is just a JSON.